spring-cloud-app-broker-security-credhub/src/main/java/org/springframework/cloud/appbroker/extensions/credentials/CredHubCredentialsGenerator.java

```diff
@@ -0,0 +1,91 @@
+
+import org.springframework.credhub.core.ReactiveCredHubOperations;
+import org.springframework.credhub.support.CredentialDetails;
+import org.springframework.credhub.support.SimpleCredentialName;
+import org.springframework.credhub.support.password.PasswordCredential;
+import org.springframework.credhub.support.password.PasswordParameters;
+import org.springframework.credhub.support.password.PasswordParametersRequest;
+import org.springframework.credhub.support.user.UserCredential;
+import org.springframework.credhub.support.user.UserParametersRequest;
+
+public class CredHubCredentialsGenerator implements CredentialGenerator {
+
+	private final ReactiveCredHubOperations credHubOperations;
+
+	public CredHubCredentialsGenerator(ReactiveCredHubOperations credHubOperations) {
+		this.credHubOperations = credHubOperations;
+	}
+
+	@Override
+	public Mono<Tuple2<String, String>> generateUser(String applicationId, String serviceInstanceId, String descriptor,
+													 int length, boolean includeUppercaseAlpha, boolean includeLowercaseAlpha,
+													 boolean includeNumeric, boolean includeSpecial) {
+		return this.credHubOperations.credentials().generate(UserParametersRequest
+			.builder()
+			.name(credentialName(applicationId, serviceInstanceId, descriptor))
+			.parameters(passwordParameters(length, includeUppercaseAlpha, includeLowercaseAlpha, includeNumeric, includeSpecial))
+			.build(), UserCredential.class)
+									 .map(CredentialDetails::getValue)
+									 .map(userCredential -> Tuples.of(userCredential.getUsername(), userCredential.getPassword()));
+	}
+
+	@Override
+	public Mono<String> generateString(String applicationId, String serviceInstanceId, String descriptor, int length,
+									   boolean includeUppercaseAlpha, boolean includeLowercaseAlpha, boolean includeNumeric,
+									   boolean includeSpecial) {
+		return credHubOperations.credentials().generate(PasswordParametersRequest
+			.builder()
+			.name(credentialName(applicationId, serviceInstanceId, descriptor))
+			.parameters(passwordParameters(length, includeUppercaseAlpha, includeLowercaseAlpha, includeNumeric, includeSpecial))
+			.build(), PasswordCredential.class)
+								.map(CredentialDetails::getValue)
+								.map(PasswordCredential::getPassword);
+	}
+
+	@Override
+	public Mono<Void> deleteUser(String applicationId, String serviceInstanceId, String descriptor) {
+		return credHubOperations.credentials().deleteByName(credentialName(applicationId, serviceInstanceId, descriptor));
+	}
+
+	@Override
+	public Mono<Void> deleteString(String applicationId, String serviceInstanceId, String descriptor) {
+		return credHubOperations.credentials().deleteByName(credentialName(applicationId, serviceInstanceId, descriptor));
+	}
+
+	private SimpleCredentialName credentialName(String applicationId, String serviceInstanceId, String descriptor) {
+		return new SimpleCredentialName(applicationId + "-" + serviceInstanceId + "-" + descriptor);
```

Comment:
This will generate a single-level name with dashes instead of the path-like structure. I think `new SimpleCredentialName(applicationId, serviceInstanceId, descriptor)` would be better.